Use binary-dep-depinfo to resolve UI test dependencies

camsteffen opened the pull request:
Closes #7343
Closes #6809
Closes #3643

changelog: none

r? @flip1995 
cc @Jarcho

Jarcho commented:
This looks like the better approach. I'll close the other one.

matthiaskrgr commented on .cargo/config:
```diff
@@ -5,4 +5,5 @@ lintcheck = "run --target-dir lintcheck/target --package lintcheck --bin lintche
 collect-metadata = "test --test dogfood --features metadata-collector-lint -- run_metadata_collection_lint --ignored"
 
 [build]
-rustflags = ["-Zunstable-options"]
+# -Zbinary-dep-depinfo allows us to track which rlib files to use for compiling UI tests
+rustflags = ["-Zunstable-options", "-Zbinary-dep-depinfo"]
```
the new "nightly-only" flag can't cause problems when running clippy tests on the stable channel inside the rustc repo in a couple of months, can it? :thinking:

flip1995 replied:
I mean, we have the nightly-only flag `-Zunstable-options` in there forever, so I assume it won't. (?)

flip1995 commented on tests/compile-test.rs:
```diff
@@ -46,50 +56,38 @@ fn third_party_crates() -> String {
+    let mut crates: HashMap<&str, &str> = HashMap::with_capacity(CRATES.len());
+    for line in depinfo.lines() {
+        let path = try {
+            let path = line.strip_suffix(':')?;
+            let part = path
+                .strip_suffix(".rlib")
+                .or_else(|| path.strip_suffix(".so"))
+                .or_else(|| path.strip_suffix(".dylib"))
+                .or_else(|| path.strip_suffix(".dll"))?;
+            let part = part.rsplit_once('-')?.0;
+            let name = part.rsplit_once(|c| matches!(c, '/' | '\\'))?.1.strip_prefix("lib")?;
+            CRATES.contains(&name).then(|| (name, path))?
```
This seems to make really strong assumptions about the format of the depinfo file. Since this is still unstable this may just change without a warning. Also reading the tracking issue https://github.com/rust-lang/rust/issues/63012, I wonder if this also works on Windows?

camsteffen replied:
Well let's spell out those assumptions (I should put this in a comment too). The depinfo file is a Makefile. For every rlib, there is a "rule" where the "target" is the rlib file and there are no "prerequisites" after the `:`. There is no extra whitespace around the `:`. On the plus side there is no assumption about where the path is in the filesystem.

I agree these are pretty strong assumptions. But I also think this buys us much better stability overall than the current solution. The file is machine-generated and seems to be unchanged for 2 years, which is quite good in this ecosystem. The flag is also used in rustc builds so that's a plus.

> I wonder if this also works on Windows?

I *think* it will since I allow for either slash and don't look at the beginning of the path. But someone should test this before merging for sure.

flip1995 replied:
> (I should put this in a comment too)

Yep :D

>  The file is machine-generated and seems to be unchanged for 2 years

Sounds good!

> I _think_ it will since I allow for either slash and don't look at the beginning of the path. But someone should test this before merging for sure.

@rust-lang/clippy anyone has access/uses a windows PC where they can test this?

Jarcho replied:
You'll need to do:

```rust
let part = part.rsplit_once('-')?.0.rsplit_once(|c| matches!(c, '/' | '\\'))?.1;
let name = part.strip_prefix("lib").unwrap_or(part);
```

dll names on windows don't have a lib prefix.

balarajupetta commented on Cargo.toml:
```diff
@@ -32,11 +32,7 @@ tempfile = { version = "3.1.0", optional = true }
-syn = { version = "1", features = ["full"] }
 # This is used by the `collect-metadata` alias.
 filetime = "0.2"
 
```
@matthiaskrgr

camsteffen commented:
* Made some stylistic changes, added comments
* Fixed depinfo parsing for Windows (@Jarcho could you test again please?)
* Added an assertion that all crates are found in depinfo
* Removed clippy_lints from the list of dependencies

flip1995 commented on tests/compile-test.rs:
```diff
@@ -24,72 +47,51 @@ fn clippy_driver_path() -> PathBuf {
+    for line in current_exe_depinfo.lines() {
+        let name_path = try {
+            // each dependency is expected to have a Makefile rule like `/path/to/crate-hash.rlib:`
+            let path = line.strip_suffix(':')?;
+            let exts = ["rlib", "so", "dylib", "dll"];
+            let part = exts.iter().find_map(|ext| path.strip_suffix(ext)?.strip_suffix('.'))?;
+            let part = part.rsplit_once('-')?.0.rsplit_once(|c| matches!(c, '/' | '\\'))?.1;
```
Have you tried doing this by parsing the lines with `PathBuf` and then manipulating it with that (get filename, remove extension)?

camsteffen replied:
Updated!

flip1995 reviewed (approved):
Thanks! I'll try this out tomorrow during sync.

Jarcho commented:
Current version is working on Windows.

flip1995 commented:
Rebased onto master

flip1995 commented:
This seemed to fix the problem in the Clippy update  rust-lang/rust#88615. I had to add `allow(unused_extern_crates)` though. Any idea why that is?

camsteffen commented:
It's `rust_2018_idioms`. Added this and deny warnings to all test modules.

flip1995 commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit 5d3fc6fc08d604889e5c52ad8b38ea474fab9a6f has been approved by `flip1995`

<!-- @bors r=flip1995 5d3fc6fc08d604889e5c52ad8b38ea474fab9a6f -->
<!-- homu: {"type":"Approved","sha":"5d3fc6fc08d604889e5c52ad8b38ea474fab9a6f","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit 5d3fc6fc08d604889e5c52ad8b38ea474fab9a6f with merge 27afd6ade4bb1123a8bf82001629b69d23d62aff...
<!-- homu: {"type":"BuildStarted","head_sha":"5d3fc6fc08d604889e5c52ad8b38ea474fab9a6f","merge_sha":"27afd6ade4bb1123a8bf82001629b69d23d62aff"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/3545200202), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/3545192981), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/3545369321)
Approved by: flip1995
Pushing 27afd6ade4bb1123a8bf82001629b69d23d62aff to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"flip1995","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/3545369321","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/3545200202","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/3545192981"},"merge_sha":"27afd6ade4bb1123a8bf82001629b69d23d62aff"} -->
